### PR TITLE
Refactor build index

### PR DESCRIPTION
**Pull request by MBkkt**
The main purpose of this PR is to make the index build operation more extensible to be able to add new types of indexes.
Commits are mostly independent and can be review one by one.

### CODE REVIEW

File: ydb/core/tx/schemeshard/schemeshard_info_types.h [hunk 1]
```diff
@@ -3235,6 +3235,41 @@ struct TIndexBuildInfo: public TSimpleRefCount<TIndexBuildInfo> {
 
 };
 
+constexpr const char* Name(TIndexBuildInfo::EState state) noexcept {
```
Comment (ijon) [inline ydb/core/tx/schemeshard/schemeshard_info_types.h:3240]: Consider moving this function to `ydb/core/tx/schemeshard/schemeshard_build_index__progress.cpp`, where is it really and only used.
Do you have practical usage for it some where else?

File: ydb/core/tx/schemeshard/schemeshard_build_index__progress.cpp [hunk 2]
```diff
@@ -640,7 +640,7 @@ struct TSchemeShard::TIndexBuilder::TTxReply: public TSchemeShard::TIndexBuilder
         case TIndexBuildInfo::EState::Filling:
         {
             // reschedule shard
-            if (buildInfo.InProgressShards.contains(shardIdx)) {
+            if (buildInfo.InProgressShards.erase(shardIdx)) {
```
Comment (ijon) [inline ydb/core/tx/schemeshard/schemeshard_build_index__progress.cpp:617]: Double erase here.

File: ydb/core/tx/datashard/build_index.cpp [hunk 3]
```diff
@@ -478,7 +478,8 @@ class TBuildScanUpload: public TActor<TBuildScanUpload<Activity>>, public NTable
             progress->Record.SetRequestSeqNoGeneration(SeqNo.Generation);
             progress->Record.SetRequestSeqNoRound(SeqNo.Round);
 
-            progress->Record.SetLastKeyAck(TSerializedCellVec::Serialize(LastUploadedKey.GetCells()));
+            // TODO(mbkkt) ReleaseBuffer isn't possible, we use LastUploadedKey for logging
+            progress->Record.SetLastKeyAck(LastUploadedKey.GetBuffer());
```
Comment (ijon) [inline ydb/core/tx/datashard/build_index.cpp:482]: Doesn't look like refactoring.
Change of data here, removal of sending responses below.
Explanation is needed.

File: ydb/core/tx/schemeshard/schemeshard__init.cpp [hunk 4]
```diff
@@ -4397,8 +4397,9 @@ struct TSchemeShard::TTxInit : public TTransactionBase<TSchemeShard> {
                     auto [it, emplaced] = Self->IndexBuilds.emplace(indexInfo->Id, indexInfo);
                     Y_ABORT_UNLESS(emplaced);
                     if (indexInfo->Uid) {
-                        std::tie(std::ignore, emplaced) = Self->IndexBuildsByUid.emplace(indexInfo->Uid, indexInfo);
-                        Y_ABORT_UNLESS(emplaced);
+                        // TODO(mbkkt) It also should be unique, but we're not sure.
+                        Y_ASSERT(!Self->IndexBuildsByUid.contains(indexInfo->Uid));
+                        Self->IndexBuildsByUid[indexInfo->Uid] = indexInfo;
```
Comment (ijon) [inline ydb/core/tx/schemeshard/schemeshard__init.cpp]: Is this change really necessary?

**Review by ijon — CHANGES_REQUESTED** [review]
Hard no to double dereferences like:
```
auto ptr = map.FindPtr(...);
auto obj = **ptr;
```

File: ydb/core/tx/schemeshard/schemeshard_build_index__progress.cpp [hunk 5]
```diff
@@ -856,14 +846,15 @@ struct TSchemeShard::TIndexBuilder::TTxReply: public TSchemeShard::TIndexBuilder
 
     bool OnNotification(TTransactionContext& txc, const TActorContext&) {
         const auto txId = CompletedTxId;
-        if (!Self->TxIdToIndexBuilds.contains(txId)) {
+        const auto* buildIdPtr = Self->TxIdToIndexBuilds.FindPtr(txId);
```
[the remark this answers is not in the document]
Comment (MBkkt) [reply ydb/core/tx/schemeshard/schemeshard_build_index__progress.cpp:843]: The point is FindPtr instead of old code, is avoid searching in hashtables twice

File: ydb/core/tx/schemeshard/schemeshard_info_types.h (hunk 1 above)
Comment (MBkkt) [reply ydb/core/tx/schemeshard/schemeshard_info_types.h:3240]: Ok

File: ydb/core/tx/datashard/build_index.cpp (hunk 3 above)
Comment (MBkkt) [reply ydb/core/tx/datashard/build_index.cpp:482]: There is no change of data here, it's same data.
Lets me explain:
LastUploadedKey -- already TSerializedCellVec (buffer + vector of cells)
Before by changes we for some reason instead of just copy buffer construct same buffer from cells.

So there is no behaviour change

File: ydb/core/tx/schemeshard/schemeshard__init.cpp (hunk 4 above)
Comment (MBkkt) [reply ydb/core/tx/schemeshard/schemeshard__init.cpp]: Before these changes there was possibility of two behavior: emplace or assign.
But I really think author of code wanted only emplace (because probably will be a lot of issues in case of duplicates), so I tried to make code more clear here.

File: ydb/core/tx/schemeshard/schemeshard_build_index__progress.cpp [hunk 6]
```diff
@@ -200,11 +200,13 @@ struct TSchemeShard::TIndexBuilder::TTxProgress: public TSchemeShard::TIndexBuil
     explicit TTxProgress(TSelf* self, TIndexBuildId id)
         : TTxBase(self, TXTYPE_PROGRESS_INDEX_BUILD)
         , BuildId(id)
-    {}
+    {
+    }
```
[the remark this answers is not in the document]
Comment (MBkkt) [reply ydb/core/tx/schemeshard/schemeshard_build_index__progress.cpp:267]: What is common practice? I'm ok to change these lines to anything.

I swear I didn't change these lines on purpose.
The main issue that we don't have something like ydb/directory wide clang-format config.

File: ydb/core/tx/schemeshard/schemeshard_import__create.cpp [hunk 7]
```diff
@@ -471,11 +471,12 @@ struct TSchemeShard::TImport::TTxProgress: public TSchemeShard::TXxport::TTxBase
...
 
         const auto uid = MakeIndexBuildUid(importInfo, itemIdx);
-        if (!Self->IndexBuildsByUid.contains(uid)) {
+        const auto* infoPtr = Self->IndexBuildsByUid.FindPtr(uid);
+        if (!infoPtr) {
             return InvalidTxId;
         }
 
-        return TTxId(ui64(Self->IndexBuildsByUid.at(uid)->Id));
+        return TTxId(ui64((**infoPtr).Id));
```
[the remark this answers is not in the document]
Comment (MBkkt) [reply ydb/core/tx/schemeshard/schemeshard_import__create.cpp:479]: > Idiomatic variant would be (*infoPtr)->.

Maybe we have different understanding of Idiomatic ;)
But ok, I can change it for you.

> So infoPtr name is a bit inaccurate.

I want to insist on such naming because it's used in all other places with FindPtr too.
And something like `infoPtrPtr` will be confused in my opinion.

File: ydb/core/tx/schemeshard/schemeshard_build_index__progress.cpp [hunk 8]
```diff
@@ -403,112 +409,104 @@ struct TSchemeShard::TIndexBuilder::TTxProgress: public TSchemeShard::TIndexBuil
...
+            } else if (!buildInfo.UnlockTxDone) {
+                Send(Self->SelfId(), MakeHolder<TEvSchemeShard::TEvNotifyTxCompletion>(ui64(buildInfo.UnlockTxId)));
             } else {
                 ChangeState(BuildId, TIndexBuildInfo::EState::Done);
                 Progress(BuildId);
             }
             break;
         case TIndexBuildInfo::EState::Done:
+        case TIndexBuildInfo::EState::Cancelled:
+        case TIndexBuildInfo::EState::Rejected:
```
[the remark this answers is not in the document]
Comment (MBkkt) [reply ydb/core/tx/schemeshard/schemeshard_build_index__progress.cpp:437]: Ok, I can change it back

File: ydb/core/tx/schemeshard/schemeshard_build_index__progress.cpp (hunk 2 above)
Comment (MBkkt) [reply ydb/core/tx/schemeshard/schemeshard_build_index__progress.cpp:617]: I don't understand where :(

File: ydb/core/tx/schemeshard/schemeshard_build_index__progress.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (MBkkt) [reply ydb/core/tx/schemeshard/schemeshard_build_index__progress.cpp:843]: `find` will return iterator, not ui64, so I don't see how it will be more convinient

File: ydb/core/tx/schemeshard/schemeshard_import__create.cpp [hunk 9]
```diff
@@ -555,14 +556,15 @@ struct TSchemeShard::TImport::TTxProgress: public TSchemeShard::TXxport::TTxBase
     }
 
     TMaybe<TString> GetIssues(TIndexBuildId indexBuildId) {
-        Y_ABORT_UNLESS(Self->IndexBuilds.contains(indexBuildId));
-        TIndexBuildInfo::TPtr indexInfo = Self->IndexBuilds.at(indexBuildId);
+        const auto* indexInfoPtr = Self->IndexBuilds.FindPtr(indexBuildId);
+        Y_ABORT_UNLESS(indexInfoPtr);
+        const auto& indexInfo = **indexInfoPtr;
```
[the remark this answers is not in the document]
Comment (MBkkt) [reply ydb/core/tx/schemeshard/schemeshard_import__create.cpp]: Changed to `*p->Get(),`

File: ydb/core/tx/schemeshard/schemeshard_build_index__progress.cpp (hunk 2 above)
Comment (MBkkt) [reply ydb/core/tx/schemeshard/schemeshard_build_index__progress.cpp:658]: Comment related to commit, it was fixed in next commits

**Review by ijon — CHANGES_REQUESTED** [review]

File: ydb/core/tx/schemeshard/schemeshard_build_index__create.cpp [hunk 10]
```diff
@@ -154,12 +155,10 @@ class TSchemeShard::TIndexBuilder::TTxCreate: public TSchemeShard::TIndexBuilder
...
                 return Reply(Ydb::StatusIds::BAD_REQUEST, explain);
             }
-        }
-
-        if (settings.has_column_build_operation()) {
-            // put some validation here for the build
-            // operation
-            buildInfo->BuildKind = TIndexBuildInfo::EBuildKind::BuildColumn;
+        } else if (settings.has_column_build_operation()) {
+            buildInfo->TargetName = settings.source_path();
```
[the remark this answers is not in the document]
Comment (MBkkt) [reply ydb/core/tx/schemeshard/schemeshard_build_index__create.cpp:159]: Because it's different for build index stage

[the remark this answers is not in the document]
Comment (MBkkt) [reply ydb/core/tx/schemeshard/schemeshard_build_index__create.cpp:159]: Filled here https://github.com/ydb-platform/ydb/blob/main/ydb/core/tx/schemeshard/schemeshard_build_index__progress.cpp#L324

File: ydb/core/tx/schemeshard/schemeshard_build_index__forget.cpp [hunk 11]
```diff
@@ -28,23 +28,22 @@ struct TSchemeShard::TIndexBuilder::TTxForget: public TSchemeShard::TIndexBuilde
...
+        if (!indexBuildInfoPtr) {
             return Reply(
                 Ydb::StatusIds::NOT_FOUND,
                 TStringBuilder() << "Index build process with id <" << indexBuildId << "> not found"
             );
         }
-
-        TIndexBuildInfo::TPtr indexBuildInfo = Self->IndexBuilds.at(indexBuildId);
-        if (indexBuildInfo->DomainPathId != domainPathId) {
+        const auto& indexBuildInfo = **indexBuildInfoPtr;
```
[the remark this answers is not in the document]
Comment (MBkkt) [reply ydb/core/tx/schemeshard/schemeshard_build_index__forget.cpp]: Fixed

**Review by ijon — APPROVED** [review]
